### PR TITLE
Add expiration logic + tests

### PR DESCRIPTION
**Pull request by 0xRusowsky**
Added a second mapping that enables the implementation of expiration logic for delegations.
This logic only applies if the expiry value is submitted. By default, there are no expirations.

### CODE REVIEW

**Comment by 0xfoobar** [comment]
Thanks! Still thinking through whether delegation expiry is a desirable feature for the complexity it adds. Certainly the delegatee has no guarantees, because the token owner can move the token at any time. This is most similar to an OS order expiring after a week, allowing users to discard hot wallets after a year or so. Thoughts @jakerockland @ryley-o?

**Comment by 0xRusowsky** [comment]
Personally, I like it, but I agree that it is not a must-have.

I think it is a nice-to-have since it may be a little bit of a burden to move funds out of a vault. The reason for this delegation standard is that owners can feel safe and"forget" about their vault because they know it can't be compromised. So this feature feels like a good value add.

File: src/DelegationRegistry.sol [hunk 1]
```diff
@@ -66,30 +90,40 @@ contract DelegationRegistry {
     * @param collection The contract address for the collection you're delegating
     * @param tokenId The token id for the token you're delegating
     * @param value Whether to enable or disable delegation for this address, true for setting and false for revoking
+    * @param expiry Optional parameter that sets the expiration date for a given delegation
```
Comment (ryley-o) [inline src/DelegationRegistry.sol:93]:
```suggestion
    * @param expiry Optional parameter that sets the expiration timestamp for a given delegation
```

File: src/DelegationRegistry.sol [hunk 2]
```diff
@@ -37,26 +42,45 @@ contract DelegationRegistry {
...
     }
 
+
     /** 
     * @notice Allow the delegate to act on your behalf for a specific NFT collection
     * @param delegate The hotwallet to act on your behalf
     * @param role The role for delegations, default is 0x0000000000000000000000000000000000000000000000000000000000000000
     * @param collection The contract address for the collection you're delegating
     * @param value Whether to enable or disable delegation for this address, true for setting and false for revoking
+    * @param expiry Optional parameter that sets the expiration date for a given delegation
```
Comment (ryley-o) [inline src/DelegationRegistry.sol:69]:
```suggestion
    * @param expiry Optional parameter that sets the expiration timestamp for a given delegation
```

File: src/DelegationRegistry.sol [hunk 3]
```diff
@@ -37,26 +42,45 @@ contract DelegationRegistry {
     * @param delegate The hotwallet to act on your behalf
     * @param role The role for delegations, default is 0x0000000000000000000000000000000000000000000000000000000000000000
     * @param value Whether to enable or disable delegation for this address, true for setting and false for revoking
+    * @param expiry Optional parameter that sets the expiration date for a given delegation
```
Comment (ryley-o) [inline src/DelegationRegistry.sol:45]:
```suggestion
    * @param expiry Optional parameter that sets the expiration timestamp for a given delegation
```

Comment (ryley-o) [reply src/DelegationRegistry.sol:45]: WDYT about also noting the default behavior if no expiry is included?

**Review by ryley-o — COMMENTED** [review]
This seems like a generally useful (and optional) feature that doesn't add complexity to the read-functions interface 😄 

Some additional gas costs to checks though.

**Comment by ryley-o** [comment]
@jakerockland - based on [this discussion](https://github.com/0xfoobar/nft-delegation/issues/5#issuecomment-1212568387), I'm assuming you would prefer not having expiry be optional here. Thoughts?

**Comment by nxt3d** [comment]
It is possible to use "value" to both include the boolean and the expiry. Will need a helper function to encode it and decode it, but would save gas I think.

**Comment by jakerockland** [comment]
> Thanks! Still thinking through whether delegation expiry is a desirable feature for the complexity it adds. Certainly the delegatee has no guarantees, because the token owner can move the token at any time. This is most similar to an OS order expiring after a week, allowing users to discard hot wallets after a year or so. Thoughts @jakerockland @ryley-o?

Yeah I have no strong opinion about how big of a value-add expiry is–I think if I really had to give a preference here I would say that I'm generally pro-expiry vs. anti-expiry, but my opinion is a weak one.

**Comment by jakerockland** [comment]
> @jakerockland - based on [this discussion](https://github.com/0xfoobar/nft-delegation/issues/5#issuecomment-1212568387), I'm assuming you would prefer not having expiry be optional here. Thoughts?

That said, I do have a strong opinion that expiry shouldn't be optional–it makes things too weird to rationalize about I think. 

I would argue that if we're adding this functionality, when setting delegation there should be an assertion that the expiry is a timestamp in the future (disallowing `0` as a valid value, making this "non-optional") and if folks want a non-expiring delegation they should have to specify so very explicitly by passing in the max uint value.

This makes this expiry behavior very explicit and something that always has to be handled intentionally by clients, vs. allowing for accidental non-expiry via passing in of the `0` value, when that may not be made as explicit by the client as it should be.

tl;dr on my opinion: we should either a) not include expiry as a feature-set, or b) if we do, it should always be explicit and non-optional

**Comment by 0xRusowsky** [comment]
@0xfoobar @nxt3d @ryley-o @jakerockland  feedback is really appreciated.

If we want expiry to be non-optional, the most simple (and gas-effective way) to do it is by merging both value and expiry. @nxt3d mentioned a helper function, but I don't think it is necessary. We only need to check that the current timestamp < expiry. If people want to delete a delegation they would only need to set expiry = 0.

If you guys agree with this, I can easily update the PR to reflect those changes

**Comment by ryley-o** [comment]
@0xRusowsky Completely agree, just replacing the mapping to a bool with a mapping to expiry is a great way to do this! 😊

**Comment by 0xRusowsky** [comment]
I closed the PR by mistake when updating the code... you can find the new PR [here](https://github.com/0xfoobar/nft-delegation/pull/11)

**Comment by jakerockland** [comment]
Yeah that SGTM @0xRusowsky ! Added one note in https://github.com/0xfoobar/nft-delegation/pull/11 re: adding an assertion to prevent accidental mis-use of setting the `expiry`.
